### PR TITLE
Use amout of snow flakes as unfreeze time indicator

### LINKED ISSUE #5527: Restore possibility to determine freeze time without freeze bar
It used to be visible via the stars. After the stars got replaced by snow flakes that information got lost. I am not saying we need to mimic the old star movement exactly. And have 1000 snow flakes on a 1000 sec freeze time. But at least reduce it to 3, 2, 1 flakes in the end.

https://user-images.githubusercontent.com/20344300/176209886-4f472082-60fe-4b46-bdcb-5bfd263933b4.mp4

### PR DESCRIPTION
**Pull request by ChillerDragon**
Use amount of snow flakes on a frozen tee to indicate the seconds till unfreeze. Only used for 2 and 1 seconds.
For longer freeze times, freeze tiles, hearts, deep freeze the old random animation is used.

closes #5527

https://github.com/ddnet/ddnet/assets/20344300/bb74aa67-1df2-450c-84c8-a3c764a39e3e

## Checklist

- [x] Tested the change ingame
- [x] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Comment by heinrich5991** [comment]
To possible other reviewers: Does this UI fit with the design of DDNet? Is it effective and usable?

**Comment by ChillerDragon** [comment]
> To possible other reviewers: Does this UI fit with the design of DDNet? Is it effective and usable?

Back when it was damage indicators instead of snow flakes it was like that. The introduction of snowflakes removed that feature.

**Comment by heinrich5991** [comment]
Yes, it was replaced by the freeze bar, which I think is well-designed.

Since this is new UI (and also about any UI), I'd like to consider the above questions.

**Comment by ChillerDragon** [comment]
>  I think is well-designed.

I strongly disagree. The freeze bar is an unacceptable design decision for me personally. Hud elements such as progress bars don't belong on entities. Makes it look like a tower defense or battle arena game.

> Since this is new UI

I would rather call it a fix than something new. The snow flakes have already been there. They were just random and pointless. Now they convey actual meaning.

**Review by Robyt3 — COMMENTED** [review]
Doesn't feel very usable to me. The snowflakes spawn with random position and size. Their contrast is rather low as they are white-transparent on a grey-white background (ninja skin). When moving slightly faster, the snowflakes do not immediately follow the tee (#5581), making it hard to follow and derive information from them. Overall, I don't think these particles are good for conveying information beyond "this tee is frozen".

The freeze bar is attached to the tee, so it can be rendered in the predicted position properly.

> They were just random and pointless. Now they convey actual meaning.

The snowflakes are a particle effect to improve the look-and-feel and I don't think they should be used to convey meaning beyond "this tee is frozen". The client has a limit for the number of particles and will simply not accept more particles when it's hit, so indicating the freeze duration would work inconsistently in the presence of many other effects like explosions and smoke.

> A top to bottom melting ice cube would be much nicer in my opinion. Something that is integrated into the game world should not be a hud element but fit the environment.

That's a cool idea. I would generally support such a design. For the last 3-5 seconds there could also be a special shaking animation to indicate exact seconds and/or an "ice-cracking" sound effect every second. I suppose that could satisfy those who want the exact one second visual/acustic timing that the freeze stars gave them (#6028).

**Comment by ChillerDragon** [comment]
I still think this is visually appealing and more correct than the current implementation. There is freeze bars and now even freeze stars for everyone who wants something more reliable. But I think it should have a nice fade-out either way.

If this pr is blocked because it's not visible enough as a timer then treat it as a visual improvement not as a timer effect.

**Comment by Robyt3** [comment]
If you want this a purely visual effect then just generate progressively less, small and/or more transparent snowflakes in the last second of freeze, instead of trying to indicate the remaining seconds exactly.

**Comment by ChillerDragon** [comment]
> If you want this as a purely visual effect then just generate progressively less, small and/or more transparent snowflakes in the last second of freeze, instead of trying to indicate the remaining seconds exactly.

What is the issue with also indicating the seconds?

**Comment by Robyt3** [comment]
> What is the issue with also indicating the seconds?

I already explained that above.

**Comment by ChillerDragon** [comment]
I don't see the reasons above as problematic thats why I asked again why these things are problematic. I personally would totally use it. I have freeze stars and bar off and count the time my self. If someone needs something more stable they can use any of the other options. This pr is not removing the freeze bar. It is just adding a new alternative. This is not a high precision timer more a visually appealing approximation. Most people will probably not notice. So its more a nice touch or easter egg. It does not have to be attached to the tee or survive the situations where there are too many particles. And it being random and hard to see is what makes it interesting and non technical and more part of the design.

Could you maybe explain me how as a result of all the reasons you named any damage is caused or how it is a downgrade to the state it was in before?

**Comment by Learath2** [comment]
I want to play with this on a bit, let me see how it looks and feels myself

**Comment by ChillerDragon** [comment]
@Learath2 rebased for you :)

**Comment by edg-l** [comment]
@ChillerDragon working on this?

**Comment by ChillerDragon** [comment]
@edg-l it always has been finished. Want me to rebase?

**Comment by edg-l** [comment]
> @edg-l it always has been finished. Want me to rebase?

hmm idk, i dont know if this is wanted, maybe ask @Learath2 to finally play around with it after this many months

**Comment by Learath2** [comment]
I did play with it for a bit throughout April. Honestly, couldn't care less, it didn't really change anything for me, probably why I even forgot to come back here and comment on it.

Is this really something anyone except @ChillerDragon wants, especially with stars being an option again?

**Comment by ChillerDragon** [comment]
@Learath2 is there anyone who does not want it?

**Comment by Learath2** [comment]
> @Learath2 is there anyone who does not want it?

@Robyt3 didn't seem to think it was a great idea. I personally didn't find it very useful. Is just you thinking it's good a good reason to merge something?

I wouldn't block it but we should still think about whether it is really a change for the better. Do we want people to start relying on something fallible? If people do start relying on this do we now have to keep this behaviour forever along with the stars?

**Comment by ChillerDragon** [comment]
Not finding it useful is imo not a good argument against it. It's a mix between design improvement and gameplay easter egg. For me it just feels more correct to fade it out visually than to have it random. It being the correct seconds is a nice touch to the 1 player that might notice it one day.


> If people do start relying on this do we now have to keep this behaviour forever along with the stars?


This is the first fair concern I hear. That of course is always a risk. But since the feature is still quite unreliable I don't think someone will actually depend on it enough to complain if it changes. But I can promise you as long as I play there will be one user demanding this feature :p. But we can also just break such features. Has happend before can happen again. I know one guy using ddnet 4 so he can see the katana in the freeze.

**Comment by heinrich5991** [comment]
I like @Robyt3's idea of improving the freeze animation. We've since re-added support for freeze stars in #8934, so I guess there's less demand for this particular feature. I think I'd prefer not to add this.

**Comment by ChillerDragon** [comment]
Pff rejecting the improvement of an animation without a proper reason is crazy

**Comment by heinrich5991** [comment]
I didn't find the reason in favor of it compelling enough. It's important not to say "yes" to every PR.

**Comment by ChillerDragon** [comment]
@heinrich5991 that statement only confirms that you have no actual reason against it. What speaks for it is that it looks better and that it has a east egg like timer. Its less cheap than just lazy randomness. If you don't need it that is okay. But why can I not get it if I want it?
